### PR TITLE
Fixes the memory leak because of project and its corresponding script info even after project is removed

### PR DESCRIPTION
**Pull request by sheetalkamat**
- fileExists and readFile of the CompilerHost created by Language service created script infos but it shouldn't because the getSourceFile is the correct api to get those since the script infos (scriptsnapshot and watcher for the file) are always associated to the project they are requested from. The issue here was that the script infos created by fileExists and readFile would get the project assigned but never removed because they weren't part of project's source file. 1bf1209   fixes this by directing those call to host instead.
- When project is removed it removed the script shot of open file/s from that project but it never use to remove it for files are not open. For big projects that meant huge strings and rest of the stuff still in memory. 2ec92b9 cleans up the ScriptInfos not associated with any project to reduce footprint left behind by closing files. This change also cleans up few pointers to make sure project, LShost, LS and other structures that are related to the project are cleaned up
- 98cb0ce  moves the cleanup of orphan script infos to after next open file so that we can reuse the script infos if the file from the previously closed project is opened right after that.

Fixes https://github.com/Microsoft/vscode/issues/28112 and #16329

### CODE REVIEW

File: src/server/editorServices.ts [hunk 1]
```diff
@@ -827,10 +827,20 @@ namespace ts.server {
                         this.assignScriptInfoToInferredProjectIfNecessary(f, /*addToListOfOpenFiles*/ false);
                     }
                 }
+
+                // Cleanup script infos that are not open and not part of any project
+                this.deleteOrphanScriptInfoNotInAnyProject();
```
Comment (mhegazy) [inline src/server/editorServices.ts]: I though we will do this in openClientFile?

Comment (sheetalkamat) [inline src/server/editorServices.ts]: This was before discussing that.. I am working on moving this to openClientFile

**Comment by nishp1** [comment]
Thanks @sheetalkamat for taking this on. Is there anyway I can help verify this bug against Code v1.13 or insider build?

File: src/server/editorServices.ts [hunk 2]
```diff
@@ -1451,6 +1484,7 @@ namespace ts.server {
             // at this point if file is the part of some configured/external project then this project should be created
             const info = this.getOrCreateScriptInfoForNormalizedPath(fileName, /*openedByClient*/ true, fileContent, scriptKind, hasMixedContent);
             this.assignScriptInfoToInferredProjectIfNecessary(info, /*addToListOfOpenFiles*/ true);
+            this.deleteOrphanScriptInfoNotInAnyProject();
```
Comment (mhegazy) [inline src/server/editorServices.ts]: please add a comment here on why we do this on open, just to make sure ppl do not wonder in the future why this is here and not in close project.

Comment (sheetalkamat) [inline src/server/editorServices.ts]: Sure

**Comment by mhegazy** [comment]
we need this in release-2.4 as well.

**Pull request merged**